onnxruntime/core/providers/cuda/tensor/gather_elements_impl.cu

```diff
@@ -143,96 +142,104 @@ __global__ void _GatherScatterElementsKernel(const T* src_data, const TIndex* in
+      OffsetCalculator<is_strided_indices>(rank, args.masked_input_strides, args.indices_fdms, args.indices_strides);  \
+  _GatherScatterElementsKernel<T, TIndex, is_gather, decltype(offset_calc), decltype(func)>                            \
+      <<<blocksPerGrid, kThreadsPerBlock, 0, stream>>>(src_data, indices_data, output_data, args.input_dim_along_axis, \
+                                                       args.input_stride_along_axis, offset_calc, func, N)
+
+#define HANDLE_GATHER_SCATTER_ELEMENTS_2D_IS_STRIDED_INDICES(src_data, is_outer_axis, is_gather) \
+  if (args.indices_strides.Size() > 0) {                                                         \
+    LAUNCH_GATHER_SCATTER_ELEMENTS_2D_KERNEL(src_data, is_outer_axis, true, is_gather);          \
+  } else {                                                                                       \
+    LAUNCH_GATHER_SCATTER_ELEMENTS_2D_KERNEL(src_data, is_outer_axis, false, is_gather);         \
+  }
 
 template <typename T, typename TIndex>
-void GatherElementsImpl(cudaStream_t stream, const int64_t rank, const int64_t axis, const T* input_data,
-                        const int64_t input_dim_along_axis, const int64_t input_stride_along_axis,
-                        const TArray<int64_t>& masked_input_strides, const TIndex* indices_data,
-                        const int64_t indices_size, const TArray<fast_divmod>& indices_fdms, T* output_data) {
-  CUDA_LONG N = static_cast<CUDA_LONG>(indices_size);
+void GatherElementsImpl(cudaStream_t stream, const T* input_data, const TIndex* indices_data, T* output_data,
+                        const GatherScatterElementsArgs& args) {
+  CUDA_LONG N = static_cast<CUDA_LONG>(args.indices_size);
   int blocksPerGrid = static_cast<int>(CeilDiv(N, kThreadsPerBlock * kThreadWorkSize));
   auto func = FuncAssignment<T>();
-  if (rank == 2) {
-    if (axis == 0) {
-      int64_t input_row_size = input_stride_along_axis;
-      LAUNCH_GATHER_SCATTER_ELEMENTS_2D_KERNEL(input_data, true, true);
+  if (args.rank == 2) {
+    int64_t input_row_size = args.masked_input_strides[0];
+    if (args.axis == 0) {
+      HANDLE_GATHER_SCATTER_ELEMENTS_2D_IS_STRIDED_INDICES(input_data, true, true);
     } else {
-      int64_t input_row_size = masked_input_strides[0];
-      LAUNCH_GATHER_SCATTER_ELEMENTS_2D_KERNEL(input_data, false, true);
+      HANDLE_GATHER_SCATTER_ELEMENTS_2D_IS_STRIDED_INDICES(input_data, false, true);
     }
     return;
   }
 
-  // Save one divmod in kernel if axis is the last dim.
-  int64_t new_rank = rank == axis + 1 ? rank - 1 : rank;
-  _GatherScatterElementsKernel<T, TIndex, true, decltype(func)><<<blocksPerGrid, kThreadsPerBlock, 0, stream>>>(
-      input_data, indices_data, output_data, new_rank, input_dim_along_axis, input_stride_along_axis,
-      masked_input_strides, indices_fdms, func, N);
+  int rank = static_cast<int>(args.rank);
+  if (args.indices_strides.Size() > 0) {
```

Comment:
shall we use a dedicate flag to indicate it is using strided slice input or not?

Reply:
Do you mean local flag or a new flag as function augment? If local, I think think it's necessary as this condition is used in only one place. If function augment, I still prefer less augments. Two or more correlated augments sometimes requires extra code to check they are compatible to each other, and also bring extra effort for the caller.